probe-rs/src/flash/flasher.rs

```diff
@@ -248,35 +227,259 @@ impl<'a> Flasher<'a> {
         Ok(r)
     }
 
-    pub fn flash_block(
-        self,
+    /// Writes a single block of data to the flash.
+    pub(super) fn flash_block(
+        &mut self,
         address: u32,
         data: &[u8],
         progress: &FlashProgress,
         do_chip_erase: bool,
         _fast_verify: bool,
-    ) -> Result<(), FlasherError> {
+    ) -> Result<(), FlashError> {
         if !self
             .region
             .range
             .contains_range(&(address..address + data.len() as u32))
         {
-            return Err(FlasherError::AddressNotInRegion {
+            return Err(FlashError::AddressNotInRegion {
                 address,
                 region: self.region.clone(),
             });
         }
 
         let mut fb = FlashBuilder::new();
         fb.add_data(address, data).expect("Add Data failed");
-        fb.program(self, do_chip_erase, true, progress)
+        self.program(&mut fb, do_chip_erase, true, false, progress)
             .expect("Add Data failed");
 
         Ok(())
     }
+
+    /// Program the contents of given `FlashBuilder` to the flash.
+    ///
+    /// If `restore_unwritten_bytes` is `true`, all bytes of a sector,
+    /// that are not to be written during flashing will be read from the flash first
+    /// and written again once the sector is erased.
+    pub(super) fn program(
+        &mut self,
+        flash_builder: &FlashBuilder,
+        mut do_chip_erase: bool,
+        restore_unwritten_bytes: bool,
+        enable_double_buffering: bool,
+        progress: &FlashProgress,
+    ) -> Result<(), FlashError> {
+        // Convert the list of flash operations into flash sectors and pages.
+        let sectors = flash_builder.build_sectors_and_pages(self, restore_unwritten_bytes)?;
+
+        let num_pages = sectors.iter().map(|s| s.pages.len()).sum();
+        let page_size = self.flash_algorithm().flash_properties.page_size;
+        let sector_size: u32 = sectors.iter().map(|s| s.size).sum();
+
+        progress.initialized(num_pages, sector_size as usize, page_size);
+
+        // Check if there is even sectors to flash.
+        if sectors.is_empty() || sectors[0].pages.is_empty() {
```

Comment:
Why do we check only the pages in the first sector here?
Should this be done for all sectors?

Reply:
Basically yes. But in this PR I didn't want to alter actual algorithm logic.